### PR TITLE
update/typescript-user

### PR DESCRIPTION
**Pull request by harshasrikara**
#### Consolidate all typescript changes into one place
 - New interface.ts file created to store global interface defintions
 - Each file now has at a minimum two interfaces for the state and props
 - Tie in and integrate all the various changes together
 - Some interfaces need to cross across files since they may define custom objects that are passed around as props --> need to ensure that all files use the interface (which should be moved into `interface.ts`) instead of `any`

### CODE REVIEW

File: src/Components/Admin/Admin.tsx [hunk 1]
```diff
@@ -111,10 +132,18 @@ class Admin extends React.Component<any, any> {
       .firestore()
       .collection("Workshop")
       .onSnapshot((snapshot) => {
-        let arr = [];
+        let arr: workshop[] = [];
         //save each workshop into an array
-        snapshot.forEach((snap) => {
-          arr.push(snap.data());
+        snapshot.forEach((snap: workshopFirebase) => {
+          let workshopObject = {
```
Comment (harshasrikara) [inline src/Components/Admin/Admin.tsx:138]: Explicit definition instead of type inference
```suggestion
          let workshopObject: workshop = {
```

File: src/Components/Admin/Admin.tsx [hunk 2]
```diff
@@ -193,10 +223,10 @@ class Admin extends React.Component<any, any> {
...
   /**
    * If workshop gets toggled between Enabled/Disabled this function gets called to update in firestore
-   * @param {*} workshopID
-   * @param {*} status
+   * @param {string} workshopID
+   * @param {number} status
    */
-  updateWorkshopStatus = (workshopID, status) => {
+  updateWorkshopStatus = (workshopID: string, status: boolean) => {
     console.log("updating status");
```
Comment (harshasrikara) [inline src/Components/Admin/Admin.tsx:230]:
```suggestion
```

File: src/Components/Admin/AdminAuth.tsx [hunk 3]
```diff
@@ -10,18 +10,29 @@ import {
...
+  password: string,
+  loginError: boolean,
+}
+
 /**
  * UI Component that contains two text fields to enter username and password as well as a submit button
  */
-class AdminAuth extends React.Component<any, any> {
+class AdminAuth extends React.Component<AdminAuthProps, AdminAuthState> {
   state = {
```
Comment (harshasrikara) [inline src/Components/Admin/AdminAuth.tsx:28]:
```suggestion
  state: AdminAuthState = {
```

File: src/Components/Admin/AdminDashboard.tsx [hunk 4]
```diff
@@ -6,14 +6,50 @@ import WorkshopEdit from "../Workshop/WorkshopEdit";
...
+  dataLoaded: boolean;
+  alert: boolean;
+  alertText: string;
+  resetAlertStatus(): void;
+}
+
+interface AdminDashboardState {
+  workshops: workshop[];
+  dataLoaded: boolean;
+  cards: any[];
```
Comment (harshasrikara) [inline src/Components/Admin/AdminDashboard.tsx:35]: There's an interface for this defined in the `CardTile.tsx` components, we need to use the same one here

File: src/Components/Layout/ConfirmationDialog.tsx [hunk 5]
```diff
@@ -8,10 +8,19 @@ import {
...
+  messageText: string;
+  workshop: any[];
+  newWorkshop: boolean;
+}
+
 /**
  * This component represents a modal which pops up when needing to confirm deletion, no other button uses this component yet
  */
-class ConfirmationDialog extends React.Component<any, any> {
+class ConfirmationDialog extends React.Component<any, ConfirmationDialogProps> {
```
Comment (harshasrikara) [inline src/Components/Layout/ConfirmationDialog.tsx:23]:
```suggestion
class ConfirmationDialog extends React.Component<ConfirmationDialogProps, any> {
```

File: src/Components/Layout/NavBar.tsx [hunk 6]
```diff
@@ -9,16 +9,33 @@ import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
...
+interface NavBarState {
+  Navlink: navlink[];
+}
+
 /**
  * This is the <NavBar /> used at the top, uses the bootstrap components but extracted it to a separate component
  */
-class NavBar extends React.Component<any, any> {
+class NavBar extends React.Component<NavBarProps, NavBarState> {
   state = {
```
Comment (harshasrikara) [inline src/Components/Layout/NavBar.tsx:29]:
```suggestion
  state: NavBarState = {
```

File: src/Components/Layout/NavBar.tsx [hunk 7]
```diff
@@ -33,7 +50,9 @@ class NavBar extends React.Component<any, any> {
   render() {
     // maps the links to be displayed at the top, eventually one of them can be like "Hello " + this.props.studentUsername
     let navlinks = this.state.Navlink.map((item, index) => (
```
Comment (harshasrikara) [inline src/Components/Layout/NavBar.tsx:52]:
```suggestion
    let navlinks = this.state.Navlink.map((item: navlink, index: number) => (
```

File: src/Components/Pages/Error404.tsx [hunk 8]
```diff
@@ -1,17 +1,19 @@
...
 import { Container } from "react-bootstrap";
-// eslint-disable-next-line spaced-comment
-// import error404 from "../../assets/404.png";
+const error404 = require("../../assets/404.png");
 
 /**
  * 404 error page, will get rendered on any path that is not defined
  */
-class Error404 extends React.Component<any, any> {
+export class Error404 extends React.Component {
```
Comment (harshasrikara) [inline src/Components/Pages/Error404.tsx:9]:
```suggestion
export class Error404 extends React.Component<any, any> {
```
Does this not require the `<any, any>` to compile?

File: src/Components/User/User.tsx [hunk 9]
```diff
@@ -3,17 +3,33 @@ import UserAuth from "./UserAuth";
 import UserDash from "./UserDash";
 import WorkshopLogin from "./WorkshopLogin";
```
Comment (harshasrikara) [inline src/Components/User/User.tsx:5]:
```suggestion
import { workshop } from "../Firebase/interface.ts"
```

File: src/Components/User/User.tsx [hunk 10]
```diff
@@ -3,17 +3,33 @@ import UserAuth from "./UserAuth";
...
-class User extends React.Component<any, any> {
-  state = {
+interface UserProps {
+  database: firebase.app.App,
+}
+
+interface UserState {
+  loggedIn: boolean,
+  workshopID: string,
+  workshop_data: any,
```
Comment (harshasrikara) [inline src/Components/User/User.tsx:13]:
```suggestion
  workshop_data: workshop,
```

File: src/Components/User/User.tsx [hunk 11]
```diff
@@ -32,7 +48,7 @@ class User extends React.Component<any, any> {
   componentDidMount() {
     this.loginListener = this.props.database
       .auth()
-      .onAuthStateChanged((user) => {
+      .onAuthStateChanged((user: any) => {
```
Comment (harshasrikara) [inline src/Components/User/User.tsx:51]:
```suggestion
      .onAuthStateChanged((user: firebase.User | any) => {
```

File: src/Components/User/User.tsx [hunk 12]
```diff
@@ -105,7 +121,7 @@ class User extends React.Component<any, any> {
           });
         }
       })
-      .catch((error) => {
+      .catch((error: firebase.firestore.FirestoreError) => {
```
Comment (harshasrikara) [inline src/Components/User/User.tsx:124]: We need this explicit type declaration for all errors in all files

File: src/Components/User/User.tsx [hunk 13]
```diff
@@ -124,20 +140,27 @@ class User extends React.Component<any, any> {
...
+      ).replace(/\./g, "%2E");
+    }
 
     //set listener on firestore
     this.progressListener = this.props.database
       .firestore()
       .collection("StudentsAtWorkshop")
       .doc(this.state.workshopID)
-      .onSnapshot((snapshot) => {
+      .onSnapshot((snapshot: firebase.firestore.DocumentData) => {
```
Comment (harshasrikara) [inline src/Components/User/User.tsx:159]: Use interface from `interface.ts`

File: src/Components/User/UserDash.tsx [hunk 14]
```diff
@@ -60,7 +85,7 @@ class UserDash extends React.Component<any, any> {
 
   // increments current level by 1. This is not their overall progress, but the stage which they are viewing.
   nextLevel = () => {
-    this.setState((state) => ({
+    this.setState((state: React.ComponentState) => ({
```
Comment (harshasrikara) [inline src/Components/User/UserDash.tsx:88]: Maybe a little too explicit here?

File: tsconfig.json [hunk 15]
```diff
@@ -2,12 +2,12 @@
   "compilerOptions": {
       "outDir": "./dist/",        // path to output directory
       "sourceMap": true,          // allow sourcemap support
-      "strictNullChecks": false,   // enable strict null checks as a best practice
```
Comment (harshasrikara) [inline tsconfig.json:5]: Why is this line deleted?

Comment (sivampatel) [reply tsconfig.json:5]: Using strict should enable strictNullChecks by default.
